semeio/jobs/correlated_observations_scaling/obs_utils.py

```diff
@@ -0,0 +1,136 @@
+import fnmatch
+from collections import namedtuple
+from copy import deepcopy
+
+from ecl.util.util import BoolVector
+from res.enkf import ActiveList, LocalObsdata, RealizationStateEnum
+
+
+def _wildcard_to_dict_list(matching_keys, entry):
+    """
+    One of either:
+    entry = {"key": "WOPT*", "index": [1,2,3]}
+    entry = {"key": "WOPT*"}
+    """
+    if "index" in entry:
+        return [{"key": key, "index": entry["index"]} for key in matching_keys]
+    else:
+        return [{"key": key} for key in matching_keys]
+
+
+def _expand_wildcard(obs_list, wildcard_key, entry):
+    """
+    Expands a wildcard
+    """
+    matching_keys = fnmatch.filter(obs_list, wildcard_key["key"])
+    return _wildcard_to_dict_list(matching_keys, entry)
+
+
+def find_and_expand_wildcards(obs_list, user_dict):
+    """
+    Loops through the user input and identifies wildcards in observation
+    names and expands them.
+    """
+    new_dict = deepcopy(user_dict)
+    for main_key, value in user_dict.items():
+        new_entries = []
+        if main_key in ("UPDATE_KEYS", "CALCULATE_KEYS"):
+            for val in value["keys"]:
+                if "*" in val["key"]:
```

Comment:
This appears to be an optimisation of `_expand_wildcard`? Should we move it into the `_expand_wildcard` function (and hence also move the knowledge of `*` being the wildcard)?

Then we could rewrite the body of the main loop as something like:
```python
    if main_key not in ("UPDATE_KEYS", "CALCULATE_KEYS"):
        continue

    new_entries = []
    for val in value["keys"]:
        new_entries.extend(_expand_wildcard(obs_list, val, val["key"]))
    new_dict[main_key]["keys"] = new_entries
```

Reply:
> Up until this point I've envisioned a clear separation between data and status reporting. Where we offer data posting and data subscription through one interface and the possibility to report your status through a different interface. I'm not completely sure why however. Maybe it is an artificial separation?

Currently, I don't value that distinction. Because to me, both (data and status) should be thought as events. And I think this is also intuitively true for many users.
 
> For the messages to the user; would this be more or less what a traditional logger interface would equip you with?

The logger interface is semantically limiting on its own. The only thing that varies is the log level. You could argue that structured logging could add the semantics needed, but then you'd have to juggle the difference between non-user records and user feedback.

Also, I think the python logger offers too many features and should not be offered to our workflow authors as a part of a _workflow api_ for getting messages back to our users. I am heavily influenced by Dave Cheney: https://dave.cheney.net/2015/11/05/lets-talk-about-logging